### ISSUE TITLE
Model is only predicting square bboxes in segmentation

### ISSUE THREAD

**Issue by pg37013** [ultralytics/ultralytics#7988]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone,

I used a pretrained model, `yolov8n-seg.pt`, to train a model to perform segmentation, and used the following configuration:

``` yaml
task: segment
mode: train
model: yolov8n-seg.pt
data: config.yaml
epochs: 200
time: null
patience: 20
batch: 64
imgsz: 512
save: true
save_period: -1
cache: false
device: 0
workers: 1
project: null
name: train22
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: true
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.2
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: false
opset: null
workspace: 4
nms: false
lr0: 0.001
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.05
hsv_s: 0.5
hsv_v: 0.2
degrees: 10
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
auto_augment: randaugment
erasing: 0.4
crop_fraction: 1.0
cfg: null
tracker: botsort.yaml
```

Although the segmentation is having the expected behavior, it seems that the model is only considering squared bounding boxes. I'm getting this:

![port_0allergy-cold-disease-flu-41284](https://github.com/ultralytics/ultralytics/assets/43340476/0e3fd29b-d682-4d17-b7a2-29ab400b8bb3)

But I want to have this instead:
![port_0allergy-cold-disease-flu-41284](https://github.com/ultralytics/ultralytics/assets/43340476/f0ca3017-44fd-4e99-88cc-9d6d444ffe28)

I already try to train a model setting `rect=True` but this didn't solve the problem. What am I doing wrong?

Can someone help me, please?

**Comment by glenn-jocher** [comment]
@pg37013 hello! It seems like you're encountering an issue where your segmentation model is predicting square bounding boxes instead of the desired shapes. This could be due to several factors, including the aspect ratio of your training data, the configuration settings, or the model's architecture itself.

Since you've already tried setting `rect=True` without success, I would suggest checking the following:

1. Ensure your dataset contains diverse aspect ratios representative of the shapes you want to detect.
2. Verify that your `config.yaml` correctly reflects your dataset's classes and annotations.
3. Consider adjusting the `overlap_mask` and `mask_ratio` parameters, as they influence the segmentation mask's quality and overlap.
4. Review the `augment` settings to ensure that the training augmentations do not constrain the bounding boxes to be square.

If the issue persists, please refer to our documentation on segmentation tasks and consider opening an issue in the Ultralytics YOLOv8 repo with detailed information about your dataset and training process for further assistance.

**Comment by pg37013** [comment]
@glenn-jocher, thanks for the quick answer.

1. I'm training using [P3M-10k dataset](https://drive.usercontent.google.com/download?id=1LqUU7BZeiq8I3i5KxApdOJ2haXm-cEv1&export=download&authuser=0), which has images with different dimensions (high resolution). However, when I trained the model, I defined `imgsz=512` and then test in a dataset with images with resolution 1080x1920, can this be the problem? Should I train the model in a high resolution?
2. Yes, my `config.yaml` file correctly reflects my dataset's classes and annotations.
3. I'm using the default parameters for `overlap_mask=True` and `mask_ratio=4`. What is the really "meaning" of mask_ratio? Should I try increase or decrease?
4. I only change the parameters `degrees=10`, `hsv_h=0.05`, `hsv_s=0.5` and `hsv_v=0.2`. However, I just realized that exists the parameter `crop_fraction: 1.0 # (float) image crop fraction for classification evaluation/inference (0-1)`. What exactly do this augmentation? The fact that it is `1.0` as default influence something related with my issue?

**Comment by glenn-jocher** [comment]
@pg37013, glad to help!

1. The discrepancy in image resolution between training and testing could indeed affect performance. Training on higher resolution images, closer to your test set resolution, might improve the model's ability to generalize and detect non-square shapes.

2. Great to hear your `config.yaml` is set up correctly.

3. The `mask_ratio` parameter influences the size of the segmentation masks relative to the bounding boxes. Adjusting it could help refine the mask shapes. A higher value may yield finer details, while a lower value might make the masks coarser.

4. The `crop_fraction` parameter is not directly related to your issue as it's more relevant for classification tasks. It doesn't influence the shape of the bounding boxes during segmentation.

Consider experimenting with the `mask_ratio` and training resolution to address the square bounding box issue. If you need more detailed guidance, our documentation on segmentation tasks can provide additional insights.

**Comment by glenn-jocher** [comment]
@pg37013 No bother at all! If the square bounding box issue persists across different image resolutions, it's less likely due to the `imgsz` parameter. Since adjusting `mask_ratio` to 1 didn't resolve the issue, it might be worth experimenting with different values or reviewing the annotations in your dataset to ensure they are accurate. If the problem continues, please provide detailed information in a new issue on our repo for further assistance. 🛠️

**Comment by pg37013** [comment]
@glenn-jocher I already check the annotations and they are accurate.

![image](https://github.com/ultralytics/ultralytics/assets/43340476/96ffc90d-82da-442c-8a25-cedd833644d7)

However, something came to my mind... Seems that the problem is related with the bounding box, but since I'm performing training for segmentation, the dataset annotations only have the annotation of the polygons, and not the bounding boxes, can this be a problem?

I already check the Dataset Guide and seems that the notations are ok, but I just want to double check.

**Comment by glenn-jocher** [comment]
@pg37013 If your annotations are accurate and you're using polygons for segmentation, the bounding boxes are typically derived from the polygons during training. The issue might not be with the annotations themselves but could be related to how the model interprets these annotations. It's possible that during the conversion from polygons to bounding boxes, the aspect ratio is not being preserved. Please ensure that the conversion process in your pipeline maintains the original aspect ratios. If the issue persists, consider opening an issue on our repo with details of your annotation format and training pipeline for further investigation. 🕵️‍♂️

**Comment by pg37013** [comment]
> It's possible that during the conversion from polygons to bounding boxes, the aspect ratio is not being preserved.

How can I check that? Can you please, give me some insights?

To train the model, I'm using the following script:

```python 
from ultralytics import YOLO

def train():

    model = YOLO('yolov8n-seg.pt')  # load a pretrained model (recommended for training)

    model.train(data='config.yaml', device=0, epochs=200, imgsz=1024, mask_ratio=10,
                    workers=1, batch=32, pretrained=True, single_cls=True, rect=True,
                    lr0=0.001, dropout=0.2, patience=20, degrees=10, hsv_h=0.05, hsv_s=0.5, hsv_v=0.2)

if __name__ == '__main__':
    train()
```
